Regional settings don't silently overwrite overlay

anothersimulacrum opened the pull request:
#### Summary
None

#### Purpose of change
Fixes https://github.com/CleverRaven/Cataclysm-DDA/issues/60394

#### Describe the solution
Track whether regional overlays have been loaded for a region, and when loading a region with that id (or any region, for "all" overlays), if any overlays have been loaded and we have not told the game to overwrite them, produce a debugmsg detailing the overlay id and what it has been loaded from.

#### Describe alternatives you've considered
This solution seems pretty clear and straightforward.

#### Testing
`tests/cata_test --mods=dda,DinoMod,tropicata '~*' ` produces an error,
`tests/cata_test --mods=dda,tropicata,DinoMod '~*' ` does not.
Load a game with dinomod and then tropicata loaded, receive an error. Exit that game without quitting the binary, and load a game with them loaded in the correct order, receive no error.

#### Additional context
This will likely cause lots of existing saves that have silently had mods loading in the wrong order to produce a message complaining about it.

anothersimulacrum committed:
Regional settings don't silently overwrite overlay

Ensure that if a regional settings overwrites a regional overlay, it
intends to do so. Otherwise, produce a debugmsg.

This will likely cause many errors in saves before this change as their
mod load order being incorrect is now producing a message.

PatrikLundell commented:
If this change results in a lot of reports for existing errors, is there a way to "repair" those saves? Getting spammed with "your save is damaged" reports with no way to recover is probably not going to be well received.

Ideally, the error report ought to indicate either how to repair a save or provide a reference to documentation that describes how to do it (in a way that most average players would be able to follow).

anothersimulacrum commented:
The way to do so it fix the loading order. I think you can do this from the main menu.

Light-Wave commented:
If all else fails, you can go to save/<world name>/mods.json and edit it to change the load order around.

anothersimulacrum commented:
TODO: Also add a feature suggesting mod ordering so that I can actually pass the all mods test :)

ZhilkinSerg commented on src/regional_settings.cpp:
```diff
@@ -448,12 +451,19 @@ static void load_region_terrain_and_furniture_settings( const JsonObject &jo,
+    if( ( regions_applied_overlay.count( new_region.id ) != 0 || loaded_all_region_overlay ) &&
+        !overwrite_overlays ) {
+        debugmsg( "Region settings for region %s (from %s) is loaded after an overlay for this region has been applied!"
+                  " Adjust world mod load order to fix.", new_region.id, src );
```
```suggestion
                  "  Adjust world mod load order to fix.", new_region.id, src );
```

anothersimulacrum committed:
Add ordering information for mods

And update the test mod list fetcher to account for it.
This allows ordering mods, as loading some mods before others could
cause errors.

Not yet recognized in game code.

LyleSY reviewed (approved):
I don't understand the code here, but the logic makes sense to me

anothersimulacrum committed:
WIP: Fix the UI

anothersimulacrum commented:
Working on the UI. Not all the way there, and I ~~want~~need tests, but here's where we are.
![Mods automatically update when selected to follow loading hints](https://user-images.githubusercontent.com/42699974/226149091-bb8b5f17-7933-41f8-b9b0-90800a395f88.webm)

Maleclypse committed:
Merge branch 'master' into regional_errorlays

Maleclypse commented:
Hi anothersimulacrum, it looks like some of the tests around region overlays fail on this Pr.
